src/main/java/com/gregtechceu/gtceu/api/recipe/ingredient/IntProviderFluidIngredient.java

```diff
@@ -0,0 +1,123 @@
+package com.gregtechceu.gtceu.api.recipe.ingredient;
+
+import com.gregtechceu.gtceu.GTCEu;
+import com.gregtechceu.gtceu.api.GTValues;
+
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.nbt.NbtOps;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.util.GsonHelper;
+import net.minecraft.util.RandomSource;
+import net.minecraft.util.valueproviders.IntProvider;
+import net.minecraft.util.valueproviders.UniformInt;
+import net.minecraftforge.fluids.FluidStack;
+
+import com.google.gson.*;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.JsonOps;
+import lombok.Getter;
+import lombok.Setter;
+import org.jetbrains.annotations.NotNull;
+
+public class IntProviderFluidIngredient extends FluidIngredient {
+
+    public static final Codec<IntProviderFluidIngredient> CODEC = ExtraCodecs.JSON
+            .xmap(IntProviderFluidIngredient::fromJson, IntProviderFluidIngredient::toJson);
+
+    @Getter
+    private final IntProvider countProvider;
+    @Setter
+    protected int sampledCount = -1;
+    @Getter
+    private final FluidIngredient inner;
+    @Setter
+    protected FluidStack[] fluidStacks = null;
+
+    protected IntProviderFluidIngredient(FluidIngredient inner, IntProvider provider) {
+        super(inner.values, provider.getMaxValue(), null);
+        this.inner = inner;
+        this.countProvider = provider;
+    }
+
+    @Override
+    public IntProviderFluidIngredient copy() {
+        IntProviderFluidIngredient ipfi = new IntProviderFluidIngredient(this, this.countProvider);
+        ipfi.setSampledCount(this.sampledCount);
+        return ipfi;
+    }
+
+    public IntProviderFluidIngredient replicate() {
+        return new IntProviderFluidIngredient(this, this.countProvider);
```

Comment:
currently this method makes the ingredient nest itself into the copy, you should do this instead:
```suggestion
        return new IntProviderFluidIngredient(this.inner, this.countProvider);
```